pinot-core/src/main/java/org/apache/pinot/core/data/manager/callback/DataManagerCallback.java

```diff
@@ -0,0 +1,106 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.manager.callback;
+
+import org.apache.pinot.core.data.manager.SegmentDataManager;
+import org.apache.pinot.spi.annotations.InterfaceStability;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.stream.MessageBatch;
+import org.apache.pinot.spi.stream.StreamPartitionMsgOffset;
+
+import java.io.IOException;
+
+/**
+ * Component inject to {@link org.apache.pinot.core.data.manager.SegmentDataManager} for handling extra logic for
+ * upsert-enabled pinot ingestion mode.
+ */
+@InterfaceStability.Evolving
+public interface DataManagerCallback {
+
+  void init() throws IOException;
+
+  /**
+   * create a {@link IndexSegmentCallback} object to allow SegmentDataManager to create proper IndexSegment that supports
+   * either append/upsert mode
+   *
+   * In append-tables callback, this method will create a DefaultIndexSegmentCallback
+   * In upsert-tables callback, this method will create a UpsertDataManagerCallbackImpl
+   *
+   * The {@link IndexSegmentCallback} will be used in the constructor of IndexSegment
+   */
+  IndexSegmentCallback getIndexSegmentCallback();
+
+  /**
+   * process the row after transformation in LLRealtimeSegmentDataManager.processStreamEvents(...) method
+   * it happens after the GenericRow has been transformed by RecordTransformer and before it is indexed by
+   * IndexSegmentImpl, to ensure we can provide other necessary data to the segment metadata.
+   *
+   * In append-tables callback, this method will do nothing
+   * In upsert-tables callback, this method will attach the offset object into the GenericRow object.
+   *
+   * The reason we need this particular logic is that in upsert table, we need to add offset data to the physical data
+   * this will help us to apply the update events from key coordinator to upsert table correctly as the offset
+   * is used as the index to identify which record's virtual column we want to update
+   *
+   * @param row the row of newly ingested and transformed data from upstream
+   * @param offset the offset of this particular pinot record
+   */
+  void processTransformedRow(GenericRow row, StreamPartitionMsgOffset offset);
+
+  /**
+   * process the row after indexing in LLRealtimeSegmentDataManager.processStreamEvents(...) method
+   * it happens after the MutableSegmentImpl has done the indexing of the current row in its physical storage
+   *
+   * In append-tables callback, this method will do nothing
```

Comment:
All realtime tables are append tables, so let us keep the terminology right. We can remove this comment, or restate it to say that the call needs to happen only if Upsert is ON for the table.